### ISSUE TITLE
False positive `map_clone` lint on coercion map.

### ISSUE THREAD

**Issue by vorner** [rust-lang/rust-clippy#2721]
Hello

I have a `Vec<String>` and I want `HashSet<&str>`, which I do with this code:

```rust
vec.iter().map(|v| -> &str { v }).collect::<HashSet<&str>>()
```

The map there is the place where rust can coerce `&String` into `&str`.

However, clippy suggests to use `.cloned` with the `map_clone` lint. This seems wrong for several reasons:

* What I do is not cloning (there's a copy of reference and a coercion)
* The proposed solution doesn't compile (because `.cloned()` still produces iterator over `&String`).

This is clippy 0.0.195.

**Comment by detrumi** [comment]
Looks like it's fixed in clippy 0.0.212, since this doesn't trigger the lint:
```
#![deny(clippy::map_clone)]

use std::collections::HashSet;

fn main() {
    let vec = vec!["a"];
    let _ = vec.iter().map(|v| -> &str { v }).collect::<HashSet<&str>>();
}
```
So I think this can be closed.
